src/tools/ceph_objectstore_tool.cc

```diff
@@ -2184,6 +2184,166 @@ int remove_clone(ObjectStore *store, coll_t coll, ghobject_t &ghobj, snapid_t cl
+	 << std::endl;
+    goto out;
+  }
+  cout << "fsid " << src->get_fsid() << std::endl;
+
+  // make sure dst is empty
+  r = dst->list_collections(collections);
+  if (r < 0) {
+    cerr << "error listing collections on dst: " << cpp_strerror(r) << std::endl;
+    goto out;
+  }
+  if (!collections.empty()) {
+    cerr << "destination store is not empty" << std::endl;
+    goto out;
+  }
+
+  r = src->list_collections(collections);
+  if (r < 0) {
+    cerr << "error listing collections on src: " << cpp_strerror(r) << std::endl;
+    goto out;
+  }
+
+  num = collections.size();
+  cout << num << " collections" << std::endl;
+  i = 1;
+  for (auto cid : collections) {
+    cout << i++ << "/" << num << " " << cid << std::endl;
+    {
+      ObjectStore::Transaction t;
+      int bits = src->collection_bits(cid);
+      if (bits < 0) {
+	cerr << "cannot get bit count for collection " << cid << ": "
+	     << cpp_strerror(bits) << std::endl;
+	goto out;
+      }
+      t.create_collection(cid, bits);
+      dst->apply_transaction(&osr, std::move(t));
+    }
+
+    ghobject_t pos;
+    uint64_t n = 0;
+    uint64_t bytes = 0, keys = 0;
+    while (true) {
+      vector<ghobject_t> ls;
+      r = src->collection_list(cid, pos, ghobject_t::get_max(), 1000, &ls, &pos);
+      if (r < 0) {
+	cerr << "collection_list on " << cid << " from " << pos << " got: "
+	     << cpp_strerror(r) << std::endl;
+	goto out;
+      }
+      if (ls.empty()) {
+	break;
+      }
+      
+      for (auto& oid : ls) {
+	//cout << "  " << cid << " " << oid << std::endl;
+	if (n % 100 == 0) {
+	  cout << "  " << std::setw(16) << n << " objects, "
+	       << std::setw(16) << bytes << " bytes, "
+	       << std::setw(16) << keys << " keys"
+	       << std::setw(1) << "\r" << std::flush;
+	}
+	n++;
+
+	ObjectStore::Transaction t;
+	t.touch(cid, oid);
+
+	map<string,bufferptr> attrs;
+	src->getattrs(cid, oid, attrs);
+	if (!attrs.empty()) {
+	  t.setattrs(cid, oid, attrs);
+	}
+
+	bufferlist bl;
+	src->read(cid, oid, 0, 0, bl);
+	if (bl.length()) {
+	  t.write(cid, oid, 0, bl.length(), bl);
+	  bytes += bl.length();
+	}
+
+	bufferlist header;
+	map<string,bufferlist> omap;
+	src->omap_get(cid, oid, &header, &omap);
+	if (header.length()) {
+	  t.omap_setheader(cid, oid, header);
+	  ++keys;
+	}
+	if (!omap.empty()) {
+	  keys += omap.size();
+	  t.omap_setkeys(cid, oid, omap);
+	}
+
+	dst->apply_transaction(&osr, std::move(t));
+      }
+    }
+    cout << "  " << std::setw(16) << n << " objects, "
+	 << std::setw(16) << bytes << " bytes, "
+	 << std::setw(16) << keys << " keys"
+	 << std::setw(1) << std::endl;
+  }
+
+  // keyring
+  cout << "keyring" << std::endl;
+  {
+    bufferlist bl;
+    string s = srcpath + "/keyring";
+    string err;
+    r = bl.read_file(s.c_str(), &err);
+    if (r < 0) {
+      cerr << "failed to copy " << s << ": " << err << std::endl;
+    } else {
+      string d = dstpath + "/keyring";
+      bl.write_file(d.c_str(), 0600);
```

Comment:
Is dstpath the right place for the keyring file when bluestone is on a block device.  Where are the files type, fsid, keyring, whoami... stored in that case?